sdk/identity/Azure.Identity/src/MsalManagedIdentityClient.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Client.AppConfig;
+
+namespace Azure.Identity
+{
+    internal class MsalManagedIdentityClient
+    {
+        private readonly AsyncLockWithValue<IManagedIdentityApplication> _clientAsyncLock;
+        private bool _isForceRefreshEnabled { get; }
+
+        protected internal bool IsSupportLoggingEnabled { get; }
+        internal ManagedIdentityId ManagedIdentityId { get; }
+        protected internal bool DisableInstanceDiscovery { get; }
+        protected internal CredentialPipeline Pipeline { get; }
+        internal Uri AuthorityHost { get; }
+
+        protected MsalManagedIdentityClient()
+        { }
+
+        public MsalManagedIdentityClient(ManagedIdentityClientOptions clientOptions)
+        {
+            // This validation is performed as a backstop. Validation in TokenCredentialOptions.AuthorityHost prevents users from explicitly
+            // setting AuthorityHost to a non TLS endpoint. However, the AuthorityHost can also be set by the AZURE_AUTHORITY_HOST environment
+            // variable rather than in code. In this case we need to validate the endpoint before we use it. However, we can't validate in
+            // CredentialPipeline as this is also used by the ManagedIdentityCredential which allows non TLS endpoints. For this reason
+            // we validate here as all other credentials will create an MSAL client.
+            Validations.ValidateAuthorityHost(clientOptions?.Options?.AuthorityHost);
+            AuthorityHost = clientOptions?.Options?.AuthorityHost ?? AzureAuthorityHosts.GetDefault();
+            IsSupportLoggingEnabled = clientOptions?.Options?.IsUnsafeSupportLoggingEnabled ?? false;
+
+            // select the correct managed identity Id.
+            if (!string.IsNullOrEmpty(clientOptions?.ClientId))
+            {
+                ManagedIdentityId = ManagedIdentityId.WithUserAssignedClientId(clientOptions.ClientId);
+            }
+            else if (clientOptions?.ResourceIdentifier != null)
```

Comment:
Is it possible for `ResourceIdentifier` to not be null, but `ToString()` to still return an empty string?

Reply:
If it can, it would be due to an invalid value, which will fail. This code assumes that if it was set to any value, it was intended to be used.